File: mobile/lib/features/reactions.dart
import 'package:flutter/material.dart';
import 'package:uuid/uuid.dart';
import '../core/api.dart';
import '../core/realtime.dart';
import '../core/preferences.dart';
import '../generated/poker.pb.dart';
import 'widgets.dart';

const reactionGlyphs = {
  'clap': '👏',
  'laugh': '😄',
  'wow': '😮',
  'angry': '😤',
  'cry': '😭',
  'nervous': '😰',
  'cold': '🥶',
  'fire': '🔥',
  'respect': '🫡',
  'sleepy': '🥱',
  'heartbeat': '🫀',
  'shark': '🦈',
  'pokerface': '😎',
  'chip': '🟠',
  'coffee': '☕',
  'clover': '🍀',
  'horseshoe': '🧲',
  'tear': '💧',
  'tomato': '🍅',
  'poop': '💩',
  'rofl': '🤣',
  'duck': '🦆',
  'turtle': '🐢',
  'knife': '🗡️',
  'flowers': '💐',
  'spotlight': '🔦',
  'crown': '👑',
  'bandage': '🩹',
  'cucumber': '🥒',
  'boomerang': '🪃',
};
const targetedReactions = {
  'chip',
  'coffee',
  'clover',
  'horseshoe',
  'tear',
  'tomato',
  'poop',
  'rofl',
  'duck',
  'turtle',
  'knife',
  'flowers',
  'spotlight',
  'crown',
  'bandage',
  'cucumber',
  'boomerang',
};

class ReactionsPanel extends StatelessWidget {
  const ReactionsPanel({super.key, required this.api, required this.realtime});
  final PokerApi api;
  final PokerRealtime realtime;
  @override
  Widget build(BuildContext context) => SafeArea(
    child: SizedBox(
      height: MediaQuery.sizeOf(context).height * .6,
      child: AsyncPanel(
        load: () => api.get('/v1.0/wallet/reaction-purchase/catalog'),
        builder: (context, result, _) => ListView(
          padding: const EdgeInsets.all(16),
          children: [
            Text(
              'Reações da mesa',
              style: Theme.of(context).textTheme.headlineSmall,
            ),
            const SizedBox(height: 16),
            Wrap(
              spacing: 8,
              runSpacing: 8,
              children: [
                for (final reaction in rows(result))
                  OutlinedButton(
                    onPressed: reaction['owned'] != true
                        ? null
                        : () async {
                            String? target;
                            final id = reaction['id'] as String;
                            if (targetedReactions.contains(id)) {
                              target = await showDialog<String>(
                                context: context,
                                builder: (context) => SimpleDialog(
                                  title: const Text('Enviar para quem?'),
                                  children: [
                                    for (final seat
                                        in realtime.snapshot?.seats ?? <Seat>[])
                                      if (seat.playerId != realtime.playerId)
                                        SimpleDialogOption(
                                          onPressed: () => Navigator.pop(
                                            context,
                                            seat.playerId,
                                          ),
                                          child: Text(seat.name),
                                        ),
                                  ],
                                ),
                              );
                              if (target == null) return;
                            }
                            realtime.command(
                              ClientMessage(
                                type: 'reaction',
                                reactionId: id,
                                targetPlayerId: target,
                                actionId: const Uuid().v4(),
                              ),
                            );
                            if (context.mounted) Navigator.pop(context);
                          },
                    child: Text(
                      '${PokerAppearance.favoriteReactions.contains(reaction['id']) ? '★ ' : ''}${reactionGlyphs[reaction['id']] ?? '☺'} ${reaction['owned'] == true ? '' : '🔒'}',
                      style: const TextStyle(fontSize: 28),
                    ),
                  ),
              ],
            ),
          ],
        ),
      ),
    ),
  );
}

File: api/internal/pokerbot/reactions.go
package pokerbot

// ReactionMoment contains only events visible to everyone at the table.
// Bots must never react to their private cards or an estimated hand strength.
type ReactionMoment int

const (
	ReactionBotWon ReactionMoment = iota
	ReactionHumanWon
	ReactionAllIn
	ReactionUncontested
)

type ReactionChoice struct {
	ID       string
	Targeted bool
}

// ChooseReaction returns a sparse, varied response to a public event. The
// actor applies the table-wide cooldown and persists the chosen reaction.
func ChooseReaction(moment ReactionMoment, random Random) (ReactionChoice, bool) {
	chance := .10
	if moment == ReactionAllIn {
		chance = .16
	}
	if random.Float64() >= chance {
		return ReactionChoice{}, false
	}
	var choices []ReactionChoice
	switch moment {
	case ReactionBotWon:
		choices = []ReactionChoice{
			{ID: "clap"}, {ID: "laugh"}, {ID: "fire"}, {ID: "shark"}, {ID: "pokerface"},
			{ID: "chip", Targeted: true}, {ID: "tear", Targeted: true},
			{ID: "tomato", Targeted: true}, {ID: "poop", Targeted: true},
			{ID: "rofl", Targeted: true}, {ID: "duck", Targeted: true},
			{ID: "spotlight", Targeted: true}, {ID: "crown", Targeted: true},
			{ID: "flowers", Targeted: true}, {ID: "bandage", Targeted: true},
		}
	case ReactionHumanWon:
		choices = []ReactionChoice{
			{ID: "clap"}, {ID: "wow"}, {ID: "angry"}, {ID: "cry"}, {ID: "respect"},
			{ID: "clover", Targeted: true}, {ID: "horseshoe", Targeted: true},
			{ID: "knife", Targeted: true}, {ID: "boomerang", Targeted: true},
		}
	case ReactionAllIn:
		choices = []ReactionChoice{
			{ID: "wow"}, {ID: "nervous"}, {ID: "heartbeat"}, {ID: "pokerface"},
			{ID: "clover", Targeted: true}, {ID: "cucumber", Targeted: true},
		}
	case ReactionUncontested:
		choices = []ReactionChoice{
			{ID: "cold"}, {ID: "sleepy"}, {ID: "pokerface"},
			{ID: "coffee", Targeted: true}, {ID: "turtle", Targeted: true},
			{ID: "cucumber", Targeted: true},
		}
	}
	if len(choices) == 0 {
		return ReactionChoice{}, false
	}
	index := int(random.Float64() * float64(len(choices)))
	if index >= len(choices) {
		index = len(choices) - 1
	}
	return choices[index], true
}
